### PR TITLE
Added prototype comparators module

### PR DESCRIPTION
**Pull request by botant**
This is a prototype of how we could provide off-the-shelf ``comparators``...

### CODE REVIEW

File: src/attr/comparators.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+       ``__le__``, ``__gt__`` and ``__ge__``.
+
+    :param bool capture_exceptions: If ``True`` (default), exceptions raised
+       within comparison methods will be captured and treated as False.
+
+    .. versionadded:: attrs-20.1.0.dev0-botant.
+    """
+
+    @attrs(slots=True, eq=False)
+    class Comparator(object):
```
Comment (jriddy) [inline src/attr/comparators.py]: Use `make_class` for this, since it already does the heavy lifting of making classes pickable:

```python
    # For pickling to work, the __module__ variable needs to be set to the
    # frame where the class is created.  Bypass this step in environments where
    # sys._getframe is not defined (Jython for example) or sys._getframe is not
    # defined for arguments greater than 0 (IronPython).
```
--- extracted from `src/attrs/_make.py:make_class`

So something like:
```python
cls = make_class('Comparator', {'value': attrib()}, slots=True, eq=False)
```

File: src/attr/comparators.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    cls = Comparator
+    cls._key = staticmethod(key)
+    cls._nonzero = staticmethod(nonzero)
+    cls.__eq__ = _make_operator("eq", "==", key)
+    cls.__ne__ = __ne__
+
+    if order:
+        cls.__lt__ = _make_operator("lt", "<", key)
+        cls = functools.total_ordering(cls)
```
Comment (jriddy) [inline src/attr/comparators.py]: `total_ordering` doesn't work for numpy arrays over .all(), since it just assumes that if `not a < b` then `a > b`, which isn't necessarily true, even in very trivial cases:

```python
>>> [1]: import numpy as np                                                                                           
...
>>> [3]: a1 = np.array([0, 1], dtype=np.int_)                                                                         
>>> [4]: a2 = np.array([1, 0], dtype=np.int_)                                                                         
>>> [5]: (a1 > a2).all()                                                                                              
<-- [5]: False
>>> [6]: (a1 < a2).all()                                                                                              
<-- [6]: False
```

That said, this could be modified to support non-total-ordering generation pretty easily

**Comment by botant** [comment]
Makes sense.  I'm not sure the overall interface via okay, though.  Feels a
little bit like throwing the kitchen sink...

**Comment by botant** [comment]
I mean that it would be nice to offer smaller bits of functionality, but
composition is very hard to achieve.

We introduce a bunch of decorators, but I don't know what performance and
backtrace would look like.

I'll write something up later today.

File: src/attr/comparators.py (hunk 2 above)
Comment (jriddy) [reply src/attr/comparators.py]: I tried to mess with this, but I'm not sure I fully understand your tests you have written on this. They seem to assume a lot about the relationship between the comparison operators, and you use total_ordering there too.

We need to unwind any assumption about a relationship between the comparison operators, because for these aggregate data types we intend to support, we can't make any of those assumptions.

I'm busy most of today, but I can take a look again tomorrow. I'm going to try to come up with some more test cases to include in this suite.

**Comment by botant** [comment]
Several things:

1) I didn't put too much thought into the tests, it was more about seeing
what it'd look like than covering the edge cases.

Feel free to disregard them entirely.

2) Great catch on numpy!

Part of the problem is that the meaning of (array_a < array_b) is
application specific, and it may not make sense at all. Is it all, any, the
determinant if the matrix is square, something else...

3) I wonder if providing comparators out-of-box won't cause more problems
than it will solve, because of the assumptions we might have to make.

Back to the drawing board.

Thanks for the review!!

**Comment by botant** [comment]
I think we are going in the wrong direction by trying to provide too much; ``attrs`` should provide a simple interface that the users can extend in a way that fit their use-case.  The more assumptions we built into this PR, the more difficult it will be to get it approved because someone somewhere will disagree, probably correctly too.

About ``numpy`` and ``pandas``, I think it is very risky to define ordering functions, and I'd rather provide ready-to-use classes that can be used for equality only, and can be extended to add ordering if the user chooses so.  The other issue with ``numpy`` and ``pandas`` is that the ``(a == b).all()`` is a bit flaky. For example, both ``numpy`` and ``pandas`` with throw a ``ValueError`` if the objects being compared have different dimensions.

Because of that, it might be easier to build something around equality functions provided in their APIs: ``numpy.array_equal(a, b)`` and for ``pandas`` ``a.equals(b)``. These two functions are *guaranteed* by them to return a boolean.

The problem is, how should we provide that? We don't want to make ``attrs`` somehow tied to  ``numpy`` and ``pandas``, so how about creating two separate projects ``attrs-numpy`` and ``attrs-pandas`` to provide comparator classes based on their respective APIs?

**Comment by jriddy** [comment]
> The more assumptions we built into this PR, the more difficult it will be to get it approved because someone somewhere will disagree, probably correctly too.

You're absolutely right. We should try to solve the general problem in a simple way that allows others to extend it to fit their needs. But that said, it's going through some of these specific examples that makes us realize what these general needs are.

I don't think making a extra projects is the right way; it seems like a lot of extra work. But think about this:

```python

@attr.s
class Example:
    arr: np.ndarray = attr.ib(eq=np.array_equal)
    df: DataFrame = attr.ib(eq=DataFrame.equals)
```

Let's put aside orderings for now and think about just getting the eq problem right, because it's hard enough. This solution seems relatively easy to me, it sidesteps the annoying naming issue, and it gets 90% of what we want without complicating the API.

**Comment by botant** [comment]
Glad you agree!

I wouldn't throw away the idea of accepting a class that will handle the
comparison, because it is generic and doesn't make any assumption. And it
is working.

The biggest question for me is what do we provide with the extended API?

So here is something in the middle:

@attr.s
class Example:
    arr: np.ndarray =
attr.ib(compare=attr.compare.using_function(eq=np.array_equal))

    df: pd.DataFrame =
attr.ib(compare=attr.compare.using_function(eq=DataFrame))

    name: str = attr.ib(compare=attr.compare.using_key(lambda s: s.lower()))


That involves no changes to we've already made to _make_eq.

We'd provide two simple functions that return comparison classes, one based
on a sort-like key, and another to which the user direct provides their own
eq method.

This also allows the user to create a full fledged compare class if that is
what they need.

I personally wouldn't like to change existing properties even to add a
feature, because attrs is so widely used... and we might create
version-related bugs, because eq=callable will evaluate to True if you
mistakenly use an old version, while the new argument `compare` or
`comparator` is safe in that regard.

Either way we agree that a modest and simple API is the way to go.

I'll add something later today so see how it looks.

I hope we can get this through in the next few days/weeks. I could use this
feature at work :-)

**Comment by botant** [comment]
Okay, I have committed a trimmed-down version of the ``comparators`` module that has two functions, ``using_key`` and ``using_functions``. I think they are quite simple and intuitive. Admittedly, the tests could be better, and I will expand/improve if you agree with the general direction.

I think I found a bug in the process of writing this, and I opened issue #626, even though it could easily be addressed in this PR...

**Comment by jriddy** [comment]
Okay I can take a look this weekend, thanks!

**Comment by botant** [comment]
Hi Josh,

Don't know if you'll even get this email, but I managed to mess up my fork
and ended up deleting the PRs in the process of fixing it. Apart from that
everything seems to be in order.

Apologies for the trouble.

**Comment by jriddy** [comment]
No trouble at all. 😉

I should have local copies of your fork, do you need anything?

**Comment by botant** [comment]
I don't think I lost anything thankfully, but annoyingly deleted the PRs
along with all comments...

The other think that happened, related or not, is that Python 2 testing
stopped working. Running tox -e py27 results in an error in hypothesis
about enum and EnumInt, and recreating the environment doesn't solve the
problem. How annoying!

File: src/attr/_make.py [hunk 3]
```diff
@@ -1189,6 +1174,23 @@ def append_hash_computation_lines(prefix, indent):
...
 
 
+def _add_method_dunders(cls, method):
+    """
+    Add __module__ and __qualname__ to *method* if possible.
+    """
+    try:
+        method.__module__ = cls.__module__
+    except AttributeError:
+        pass
```
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/_make.py]: I don't know to be honest. This function was extracted from ClassBuilder and I didn't change it. 

I suppose that at some point there was the intention to set ``__qualname__`` and ``__module__`` but these properties weren't part of the class being modified. Maybe this is a Python2/3 thing?

I agree that a little comment explaining why this was being ignored would have made the function more palatable.

File: src/attr/comparators.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    .. versionadded:: attrs-20.1.0.dev0-botant.
+    """
+
+    def _create_method(func):
+        def __func__(self, other):
+            """
+            Automatically created by attrs.
+            """
+            if other.__class__ is self.__class__:
```
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: Absolutely, thanks catching it.  Same thing also applies to the other ``using_`` function.

File: src/attr/comparators.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                return func(self.value, other.value)
```
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: You can drop ``eq=True`` and make it a bit shorter :-)

I would be nice if there was a way to imply ``eq`` and ``order`` from the comparator being used. Maybe should go back to your idea and use ``comparator`` to override ``eq`` and ``order``, and raise an exception if they are used in conjunction. What do you think?

File: src/attr/comparators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: About ``if other.__class__ is self.__class__``: should we enforce the view that different classes cannot be compared? I understand (and accept) that this view might have been a necessary evil introduced in ``attrs`` in order for it to handle values of any type, but I'm not sure we should do this here.

Rationale for binning the ``if``: if the user is providing custom comparison functions, then the user can also decide whether it is okay to compare different classes.

Rationale for keeping the ``if``: it is in keeping with ``attrs`` behaviour, and it may simplify things for the users a little bit.

How about adding an argument to ``using_key`` and ``using_functions`` and let the user decide how the ``comparator`` should behave? Something like ``require_same_class=True``?

File: src/attr/comparators.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: To test this out, I have committed a bit a new function to check ``cmp``, ``eq``, ``order`` and ``comparator`` for ``attribs``.  It is annoying that it is not possible to tell whether the user wanted ``order=True`` or not by looking at the ``comparator`` because it will always have the dunder methods. We can add some sort of *protocol* method, but I think this is not worth the hassle.

At the moment, all my function can do is say *if a comparator is present, then eq=True*.

File: src/attr/comparators.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: I have added ``require_same_class=True`` to both functions...

[the remark this answers is not in the document]
Comment (botant) [reply src/attr/comparators.py]: ... and fixed the bug you found earlier.

**Comment by jriddy** [comment]
Hey sorry, I haven't been paying attention here, I've been busy with buying a house and changing jobs...I'll take a look at this this weekend and try to invest more time in this (even during the week now :wink: )

**Comment by botant** [comment]
Really?  I'm actually trying to buy my first house as well, having to deal with mortgage brokers, solicitors and all that stuff for the first time...  but no job change hopefully.  Good luck on both!

**Comment by jriddy** [comment]
Haha both are done now, but I have to plan out a series of renovations to
get the place ready, so my weekends will be busy, but my weekdays will have
more free time.

Good on you using a mortgage broker, they can really help navigate the
financial side.
